[documentation] How to debug "Connection prematurely closed BEFORE response" (#1905)
reactor/reactor-netty: docs/asciidoc/faq.adoc

Reactor Netty adds information for the connection at the beginning of every log record (when this is possible).
Use the id of the connection in order to extract all log records for a particular HTTP request.
For more information see <<faq.logging-prefix>>

[[faq.connection-closed]]
== How can I debug "Connection prematurely closed BEFORE response"?

> By default, Reactor Netty clients use a connection pooling. When a connection is acquired from the connection pool, it is checked whether it is still opened.

Reviewer: ```suggestion
By default, Reactor Netty clients uses connection pooling. When a connection is acquired from the connection pool, it is checked to be open still.
```

> By default, Reactor Netty clients use a connection pooling. When a connection is acquired from the connection pool, it is checked whether it is still opened.

Reviewer 2: Change to:

By default, Reactor Netty clients use connection pooling. When a connection is acquired from the connection pool, it is checked to see whether it is still open.

However, the connection can be closed at any time after the acquisition. There are many reasons that can cause connection to be closed.

> In most cases the setup might not be like: the client directly sends to the server, but there might be many other network components between them: proxies, load balancers etc.

Reviewer: the `setup might not be like:` part of this sentence is unclear. not like what? there are two `:` in this sentence, which makes me think it was meant to be two separate sentences?

Reviewer 2: There are two related ideas here. I would write it as two sentences:

In most cases, the client might not send directly to the server. Instead, there might be other network components (proxies, load balancers, and so on) between them.

> If on the client side you observe `Connection prematurely closed BEFORE response`, perform the checks below to identify the reason for the connection close:

Reviewer 2: Change to:

If, on the client side, you observe `Connection prematurely closed BEFORE response`, perform the following checks to identify the reason for the connection being closed:

> * Obtain https://en.wikipedia.org/wiki/Tcpdump[tcp dump] and check which peer sends `FIN`/`RST` signal

Reviewer 2: Change to:

* Obtain a https://en.wikipedia.org/wiki/Tcpdump[TCP dump] and check which peer sends a `FIN`/`RST` signal.

> * Check your network connection

Reviewer 2: Add a period at the end. (Because each of these is a complete sentence, they require periods. One of the items having multiple sentences also forces the others to be sentences and have periods.)

> * Check your Firewall/VPN

Reviewer 2: Change to:

* Check your Firewall and VPN.

> * Check for any proxies and load balancers. Do they have some kind of idle timeout configuration? (the connection is closed when there is no incoming data for a certain period of time)

Reviewer 2: Move the question mark to being last (after the parenthetical content).

> * Check the target server. Are there configurations related to:

Reviewer 2: Change to:

Check the target server. Are there configurations related to any of the following?

** idle timeout (the connection is closed when there is no incoming data for a certain period of time)
** limit for buffering data in memory
** multipart exceeds the max file size limit
** bad request

> Consider checking <<Timeout Configuration>>. The section describes various timeout configuration options available for Reactor Netty clients.

Reviewer 2: Add "that are" between "options" and "available".
